stratum/hal/lib/tdi/dpdk/dpdk_add_subtree_interface.cc

```diff
@@ -0,0 +1,323 @@
+// Copyright 2018 Google LLC
+// Copyright 2018-present Open Networking Foundation
+// Copyright 2022 Intel Corporation
+// SPDX-License-Identifier: Apache-2.0
+
+// Implements the DPDK-specific YangParseTreePaths::AddSubtreeInterface() method.
+// The supporting functions are in other files.
+
+#include "stratum/hal/lib/yang/yang_parse_tree_paths.h"
+
+#include <string>
+
+#include "stratum/hal/lib/common/common.pb.h"
+#include "stratum/hal/lib/common/gnmi_publisher.h"
+#include "stratum/hal/lib/tdi/dpdk/dpdk_parse_tree_interface.h"
+#include "stratum/hal/lib/yang/yang_parse_tree.h"
+#include "stratum/hal/lib/yang/yang_parse_tree_interface.h"
+
+namespace stratum {
+namespace hal {
+
+using namespace ::stratum::hal::yang::interface;
+
+// Paths of leaves created by this method are defined manually by analysing
+// existing YANG model files. They are hard-coded, and, as the YANG language
+// does not provide a means to express leaves' semantics, their mapping to code
+// implementing their function is also done manually.
+// TODO(b/70300012): Implement a tool that will help to generate this code.
+
+TreeNode* YangParseTreePaths::AddSubtreeInterface(
+    const std::string& name, uint64 node_id, uint32 port_id,
+    const NodeConfigParams& node_config, YangParseTree* tree) {
+  // No need to lock the mutex - it is locked by method calling this one.
+  uint64 mac_address = kDummyMacAddress;
+
+  TreeNode* node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("last-change")());
+  SetUpInterfacesInterfaceStateLastChange(node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("config")("ifindex")());
+  SetUpInterfacesInterfaceStateIfindex(node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("name")());
+  SetUpInterfacesInterfaceStateName(name, node);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("oper-status")());
+  SetUpInterfacesInterfaceStateOperStatus(node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("admin-status")());
+  SetUpInterfacesInterfaceStateAdminStatus(node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("loopback-mode")());
+  SetUpInterfacesInterfaceStateLoopbackMode(node_id, port_id, node, tree);
+
+  node = tree->AddNode(
+      GetPath("interfaces")("virtual-interface", name)("state")("hardware-port")());
+  SetUpInterfacesInterfaceStateHardwarePort(name, node, tree);
```

Comment:
hardware-port, loopback-mode, port-speed, negotiated-port-speed - Not applicable to DPDK